### ISSUE TITLE
Server closing connection to BotFarm

### ISSUE THREAD

**Issue by Trystanosaurus** [jackpoz/BotFarm#42]
I'm trying to run bot farm but haven't managed to get it to create any characters.

I've made an account with create account privileges and tested in game that that works. This bit of it seems to be going fine. However it's not getting as far as creating characters.

A typical log is below, it seems to log in fine, then it says there are no characters and the server closes the connection. This happens with RandomBots set to true or false.

After the first few shutdown connections I start getting errors saying "The seqence contains no matching element    in System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   in BotFarm.BotFactory.RemoveBot(BotGame bot)
   in Client.World.Network.WorldSocket.HandlePacket(InPacket packet)"

My trinity game log is also logging "WorldSocket::HandleAuthSession: Authentication failed for account: 17 ('BOT1240406805') address: 192.168.1.64"

This is with the current build of Trinity Core (3.3.5)

Starting BotFactory
Waiting for BotFactory account to login
Setting up bot factory with 10 bots
BOT1240406805 - Connecting to realmlist... 
BOT443835549 - Connecting to realmlist... 
BOT1240406805 - Connecting to realmlist... 
BOT904510679 - Connecting to realmlist... 
BOT1175610644 - Connecting to realmlist... 
BOT969339491 - Connecting to realmlist... 
BOT904510679 - Connecting to realmlist... 
BOT1175610644 - Connecting to realmlist... 
BOT1240406805 - Connecting to realmlist... 
BOT200893845 - Connecting to realmlist... 
Finished setting up bot factory with 10 bots in 00:00:00.0215997
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT443835549 - Authentication succeeded!
BOT443835549 - Requesting realm list
BOT904510679 - Authentication succeeded!
BOT904510679 - Requesting realm list
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1175610644 - Authentication succeeded!
BOT1175610644 - Requesting realm list
BOT969339491 - Authentication succeeded!
BOT969339491 - Requesting realm list
BOT904510679 - Authentication succeeded!
BOT904510679 - Requesting realm list
BOT1175610644 - Authentication succeeded!
BOT1175610644 - Requesting realm list
BOT1240406805 - Connecting to realm Trinity
BOT443835549 - Connecting to realm Trinity
BOT904510679 - Connecting to realm Trinity
BOT1240406805 - Connecting to realm Trinity
BOT1175610644 - Connecting to realm Trinity
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT200893845 - Authentication succeeded!
BOT200893845 - Requesting realm list
BOT969339491 - Connecting to realm Trinity
BOT904510679 - Connecting to realm Trinity
BOT1175610644 - Connecting to realm Trinity
BOT1240406805 - Connecting to realm Trinity
BOT200893845 - Connecting to realm Trinity
BOT1175610644 - Server has closed the connection
BOT1240406805 - Server has closed the connection
BOT1240406805 - Server has closed the connection
BOT904510679 - Server has closed the connection
BOT1175610644 - Connecting to realmlist... 
BOT904510679 - Connecting to realmlist... 
BOT1240406805 - Connecting to realmlist... 
BOT1240406805 - Connecting to realmlist... 
BOT1175610644 - Authentication succeeded!
BOT1175610644 - Requesting realm list
BOT904510679 - Authentication succeeded!
BOT904510679 - Requesting realm list
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1175610644 - Connecting to realm Trinity
BOT904510679 - Connecting to realm Trinity
BOT1240406805 - Connecting to realm Trinity
BOT1240406805 - Connecting to realm Trinity
BOT443835549 - No characters found!
BOT443835549 - Removing current bot because there are no characters
BOT443835549 - Server has closed the connection
BOT904510679 - No characters found!
BOT904510679 - Removing current bot because there are no characters
BOT904510679 - Server has closed the connection
BOT969339491 - No characters found!
BOT969339491 - Removing current bot because there are no characters
BOT969339491 - Server has closed the connection
BOT1175610644 - No characters found!
BOT1175610644 - Removing current bot because there are no characters
BOT1175610644 - Server has closed the connection
BOT1240406805 - No characters found!
BOT1240406805 - Removing current bot because there are no characters
BOT1240406805 - Server has closed the connection
BOT200893845 - No characters found!
BOT200893845 - Removing current bot because there are no characters
BOT200893845 - Server has closed the connection
BOT1240406805 - Server has closed the connection
BOT1240406805 - Connecting to realmlist... 
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1240406805 - Connecting to realm Trinity
BOT904510679 - No characters found!
BOT904510679 - Removing current bot because there are no characters
BOT904510679 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1175610644 - No characters found!
BOT1175610644 - Removing current bot because there are no characters
BOT1175610644 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1240406805 - No characters found!
BOT1240406805 - Removing current bot because there are no characters
BOT1240406805 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1240406805 - Server has closed the connection
BOT1240406805 - Connecting to realmlist... 
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1240406805 - Connecting to realm Trinity
BOT1240406805 - No characters found!
BOT1240406805 - Removing current bot because there are no characters
BOT1240406805 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1240406805 - Server has closed the connection
BOT1240406805 - Connecting to realmlist... 
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1240406805 - Connecting to realm Trinity
BOT1240406805 - No characters found!
BOT1240406805 - Removing current bot because there are no characters
BOT1240406805 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1240406805 - Server has closed the connection
BOT1240406805 - Connecting to realmlist... 
BOT1240406805 - Authentication succeeded!
BOT1240406805 - Requesting realm list
BOT1240406805 - Connecting to realm Trinity
BOT1240406805 - No characters found!
BOT1240406805 - Removing current bot because there are no characters
BOT1240406805 - Sekvensen innehåller inga matchande element    vid System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   vid BotFarm.BotFactory.RemoveBot(BotGame bot)
   vid Client.World.Network.WorldSocket.HandlePacket(InPacket packet)
BOT1240406805 - Server has closed the connection

**Comment by jackpoz** [comment]
It might be easier to try with just 1 single bot the first time

**Comment by Trystanosaurus** [comment]
Thanks for the quick reply. I've tried with one and it's still having problems, it seems to be because there are no characters on the bot accounts, do I have to create those manually?

When I type stats after it's finished it shows 0 active, 0 connected, 0 ingame.

Starting BotFactory
Waiting for BotFactory account to login
Setting up bot factory with 1 bots
BOT455709970 - Connecting to realmlist... 
Finished setting up bot factory with 1 bots in 00:00:00.0113512
BOT455709970 - Authentication succeeded!
BOT455709970 - Requesting realm list
BOT455709970 - Connecting to realm Trinity
BOT455709970 - No characters found!
BOT455709970 - Removing current bot because there are no characters
BOT455709970 - Server has closed the connection

**Comment by jackpoz** [comment]
Please delete the botsinfos.xml file to allow it to create a fresh new account with a fresh new character.
The current behavior creates a new character only when creating the account, I'll have to change it to create a new character if no characters have been found.

**Comment by Trystanosaurus** [comment]
Thanks, I've removed the botsinfos.xml and am now getting an error where it looks to be failing to log in the new accounts. I've checked the database and the account is there but has never logged in. Restarting BotFarm logs the account in but then it's back to the problem of having no characters.

I hope this helps, let me know if there's anything else I can provide.

Log is below:
Starting BotFactory
Waiting for BotFactory account to login
Setting up bot factory with 1 bots
Creating new bot
BOT722977947 - Connecting to realmlist... 
Finished setting up bot factory with 1 bots in 00:00:00.0169382
BOT722977947 - Unknown account name
Shutting down BotFactory

Stats is showing:
Username: BOT722977947
    Behaviour: Default
    Running: True
    Connected: False
    Logged In: False
    Position: Invalid
    Group Leader: Not in group
    Last Received Packet: LOGON_CHALLENGE
    Last Sent Packet: LOGON_CHALLENGE
    Last Update() call: 09:14:53
1 bots are active
0 bots are connected
0 bots are ingame
